sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,766 @@
+```
+
+### Emulation of specialization constants
+
+Emulation of specialization constants is performed by converting them into
+kernel arguments.
+
+Overall idea is that DPC++ runtimes packs all specialization constants into a
+single buffer, which is passed as an extra implicit kernel argument. Then the
+compiler instead of lowering `__sycl_get*2020SpecConstantValue` intrinsics into
+SPIR-V friendly IR replaces it with extracting an element from that buffer.
+
+"All" specialization constants here means complete list of specialization
+constants encountered in an application or a shared library which is being
+compiled: that list is computed by `sycl-post-link` tool and communicated to
+the runtime through device image properties like it is described in "Support for
+native specialization constants" section.
+
+#### DPC++ Headers
+
+The same DPC++ Headers are used for native and emulated specialization constants
+and their design is decribed in the corresponding sub-section of "Support for
+native specialization constants" section.
+
+However, that part of the document doesn't describe the third argument of
+`__sycl_get*2020SpecConstantValue` intrinsics: it is a pointer to a runtime
+buffer, which holds values of all specialization constants and should be used
+to retrieve their values in device code.
+
+This pointer is stored within `kernel_handler` object and it is initialized only
+if our target doesn't support native specialization constants.
+Since `kernel_handler` object is not captured by SYCL kernel funtion, it means
+that we are not able to employ some header-only solution here and need help of
+the compiler.
+
+DPC++ FE searches for functions marked with `sycl_kernel` attribute to handle
+them and turn into entry points of device code.
+
+
+#### DPC++ FE
+
+DPC++ FE should look for `kernel_handler` argument in a function marked with
+`sycl_kernel" attribute. If such argument is present, it means that this kernel
+can access specialization constants and therefore FE needs to do the following:
+
+If native specialization constants are supported:
+- create `kernel_handler` object
+- use default constructor to initialize it
+- pass that `kernel_handler` object to user-provided SYCL kernel function
+
+If native specialization constants are not supported:
+- generate one more kernel argument for passing a buffer with specialization
+  constants values
+- create `kernel_handler` object
+- initialize that `kernel_handler` object with newly created kernel argument
+- pass that `kernel_handler` object to user-provided SYCL kernel function
+- Provide information about new kernel argument through the integration header
+
+So, having the following as the input:
+```
+template <typename KernelName, typename KernelType>
+__attribute__((sycl_kernel)) void
+kernel_single_task(const KernelType &KernelFunc, kernel_handler kh) {
+  KernelFunc(kh);
+}
+```
+For the target which has native support for specialization constatns DPC++ FE
```

Comment:
Do you mean target which does not have native support ?

Reply:
Right, fixed.